charmbits/httpservice/httpservice_test.go

```diff
@@ -0,0 +1,321 @@
+	expectPanic:   `setFieldValue function argument to RegisterRelationType has wrong return type, got int, want error`,
+}, {
+	about:         "too few args",
+	setFieldValue: func(*struct{}) error { panic("unreachable") },
+	expectPanic:   `setFieldValue function argument to RegisterRelationType has wrong argument count; got 1, want 2`,
+}, {
+	about:         "too many args",
+	setFieldValue: func(*struct{}, int, int) error { panic("unreachable") },
+	expectPanic:   `setFieldValue function argument to RegisterRelationType has wrong argument count; got 3, want 2`,
+}, {
+	about:         "first arg not pointer",
+	setFieldValue: func(struct{}, int) error { panic("unreachable") },
+	expectPanic:   `setFieldValue function argument to RegisterRelationType has wrong first argument type; got struct {} want pointer`,
+}}
+
+func (s *suite) TestRegisterRelation(c *gc.C) {
+	for i, test := range registerRelationTests {
+		c.Logf("test %d: %s", i, test.about)
+		httpservice.ClearRegisteredRelations()
+		if test.expectPanic != "" {
+			c.Assert(func() {
+				httpservice.RegisterRelationType(nopRegisterField, test.setFieldValue)
+			}, gc.PanicMatches, "cannot register relation: "+test.expectPanic)
+			continue
+		}
+		httpservice.RegisterRelationType(nopRegisterField, test.setFieldValue)
+		t, d, ok := httpservice.RegisteredRelationInfo(test.expectT)
+		c.Assert(ok, gc.Equals, true)
+		c.Assert(t, gc.Equals, test.expectT)
+		c.Assert(d, gc.Equals, test.expectD)
+	}
+}
+
+var registerTests = []struct {
+	about            string
+	registerField    func(r *hook.Registry, tag string) (getVal func() (interface{}, error))
+	setFieldValue    interface{}
+	serviceName      string
+	httpRelationName string
+	handler          interface{}
+	expectRelations  map[string]charm.Relation
+}{{
+	about: "no panic, simple relation",
+	registerField: func(r *hook.Registry, tag string) func() (interface{}, error) {
+		if tag != "foo-tag" {
+			panic("unexpected tag " + tag)
+		}
+		r.RegisterRelation(charm.Relation{
+			Name:      "foo",
+			Interface: "foo-interface",
+			Role:      charm.RoleRequirer,
+			Scope:     charm.ScopeGlobal,
+		})
+		return func() (interface{}, error) {
+			panic("unreachable")
+		}
+	},
+	setFieldValue: func(f *string, v string) error {
+		*f = "relation val: " + v
+		return nil
+	},
+	serviceName:      "fooservice",
+	httpRelationName: "httpserver",
+	handler: func(s string, r *struct {
+		Foo string `httpservice:"foo-tag"`
+	}) (httpservice.Handler, error) {
+		return nil, fmt.Errorf("no handler yet")
+	},
+	expectRelations: map[string]charm.Relation{
+		"foo": {
+			Name:      "foo",
+			Role:      charm.RoleRequirer,
+			Interface: "foo-interface",
+			Limit:     1,
+			Scope:     charm.ScopeGlobal,
+		},
+		"httpserver": {
+			Name:      "httpserver",
+			Role:      charm.RoleProvider,
+			Interface: "http",
+			Scope:     charm.ScopeGlobal,
+		},
+	},
+},
+
+// TODO test all error cases in newHandlerInfo
```

Comment:
All TODOs in this project are implicitly TODO(rog)...
I may do an automatic edit to change them to say that explicitly at
some point in the future.